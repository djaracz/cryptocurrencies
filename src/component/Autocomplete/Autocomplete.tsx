import { SyntheticEvent, useCallback } from "react";
import {
  Autocomplete as MUIAutocomplete,
  AutocompleteValue,
  TextField,
  TextFieldProps,
} from "@mui/material";

export interface AutocompleteProps<O> {
  options: O[];
  disabled: boolean;
  values: AutocompleteValue<O, true, false, true>;
  textFieldProps: Partial<TextFieldProps>;
  getOptionLabel: (option: O) => string;
  onChange: (selectedOptions: O[]) => void;
}

export const Autocomplete = <O,>({
  options,
  disabled,
  values,
  textFieldProps,
  getOptionLabel,
  onChange,
}: AutocompleteProps<O>) => {
  const handleOnChange = useCallback(
    (e: SyntheticEvent, values: AutocompleteValue<O, true, false, true>) =>
      onChange(values as O[]),
    [onChange]
  );

  return (
    <MUIAutocomplete
      multiple
      options={options}
      getOptionLabel={getOptionLabel}
      sx={{ width: "100%" }}
      disabled={disabled}
      onChange={handleOnChange}
      value={values as any}
      renderTags={(options) =>
        options.reduce(
          (acc, o) =>
            acc ? `${acc}, ${getOptionLabel(o)}` : getOptionLabel(o),
          ""
        )
      }
      renderInput={(params) => (
        <TextField {...params} variant="outlined" {...textFieldProps} />
      )}
    />
  );
};
