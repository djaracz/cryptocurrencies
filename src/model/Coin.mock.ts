import { Coin } from "./Coin";

export const coins: Coin[] = [
  {
    ath: 69045,
    ath_change_percentage: -70.58616,
    ath_date: "2021-11-10T14:24:11.849Z",
    atl: 67.81,
    atl_change_percentage: 29849.88065,
    atl_date: "2013-07-06T00:00:00.000Z",
    circulating_supply: 19137543,
    current_price: 20344,
    high_24h: 20439,
    id: "bitcoin",
    image:
      "https://assets.coingecko.com/coins/images/1/large/bitcoin.png?1547033579",
    last_updated: "2022-08-31T12:05:02.637Z",
    low_24h: 19590.92,
    market_cap: 389106035937,
    market_cap_change_24h: -1074243445.8865967,
    market_cap_change_percentage_24h: -0.27532,
    market_cap_rank: 1,
    max_supply: 21000000,
    name: "Bitcoin",
    price_change_24h: -58.51002852483725,
    price_change_percentage_24h: -0.28678,
    symbol: "btc",
    total_supply: 21000000,
    total_volume: 32340055570,
  },
  {
    ath: 4878.26,
    ath_change_percentage: -67.33692,
    ath_date: "2021-11-10T14:24:19.604Z",
    atl: 0.432979,
    atl_change_percentage: 367906.45469,
    atl_date: "2015-10-20T00:00:00.000Z",
    circulating_supply: 120325702.869246,
    current_price: 1597.5,
    high_24h: 1614.12,
    id: "ethereum",
    image:
      "https://assets.coingecko.com/coins/images/279/large/ethereum.png?1595348880",
    last_updated: "2022-08-31T12:04:25.855Z",
    low_24h: 1479.78,
    market_cap: 192011349748,
    market_cap_change_24h: 1418488290,
    market_cap_change_percentage_24h: 0.74425,
    market_cap_rank: 2,
    max_supply: 21000000,
    name: "Ethereum",
    price_change_24h: 9.92,
    price_change_percentage_24h: 0.62466,
    symbol: "eth",
    total_supply: 120325702.869246,
    total_volume: 21797779943,
  },
];
