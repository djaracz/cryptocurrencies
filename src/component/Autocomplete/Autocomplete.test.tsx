import { render, screen, fireEvent, prettyDOM } from "@testing-library/react";
import { Coin } from "../../model/Coin";
import { coins } from "../../model/Coin.mock";
import { Autocomplete, AutocompleteProps } from "./Autocomplete";

let onChange: () => void;

beforeEach(() => {
  onChange = jest.fn();
});

const renderComponent = (partialProps: Partial<AutocompleteProps<Coin>>) => {
  return render(
    <Autocomplete
      textFieldProps={{
        label: "Label",
        placeholder: "Placeholder",
      }}
      disabled={false}
      getOptionLabel={(c) => c.name}
      onChange={onChange}
      options={coins}
      values={[]}
      {...partialProps}
    />
  );
};

test("should render properly", () => {
  const { container } = renderComponent({});
  expect(container).toMatchSnapshot();
});

test("should render with selected values", () => {
  const { container } = renderComponent({
    values: [coins[0]],
  });
  expect(container).toMatchSnapshot();
});

test("should open selector on open icon click", () => {
  renderComponent({});

  const openBtn = screen.getAllByTitle("Open");
  expect(openBtn).toBeDefined();

  fireEvent.click(openBtn[0]);
  const closeBtn = screen.getAllByTitle("Close");
  expect(closeBtn).toBeDefined();
});

test("should call onChange event on clear options click", () => {
  renderComponent({
    values: [coins[0]],
  });

  const clearBtn = screen.getByTitle("Clear");
  expect(clearBtn).toBeDefined();
  expect(onChange).toBeCalledTimes(0);

  fireEvent.click(clearBtn);
  expect(onChange).toBeCalledTimes(1);
});
