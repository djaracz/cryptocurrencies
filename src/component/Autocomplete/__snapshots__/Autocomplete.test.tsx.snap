// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<div>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-oarg8b-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="false"
        for=":r0:"
        id=":r0:-label"
      >
        Label
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r0:"
          placeholder="Placeholder"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1ftyaf0"
          >
            <span>
              Label
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render with selected values 1`] = `
<div>
  <div
    class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-oarg8b-MuiAutocomplete-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for=":r2:"
        id=":r2:-label"
      >
        Label
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gywuxd-MuiInputBase-root-MuiOutlinedInput-root"
      >
        Bitcoin
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          aria-invalid="false"
          autocapitalize="none"
          autocomplete="off"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-152mnda-MuiInputBase-input-MuiOutlinedInput-input"
          id=":r2:"
          placeholder="Placeholder"
          role="combobox"
          spellcheck="false"
          type="text"
          value=""
        />
        <div
          class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Clear"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
            tabindex="-1"
            title="Clear"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Open"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            tabindex="-1"
            title="Open"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1in441m"
          >
            <span>
              Label
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
